src/EnergyPlus/api/runtime.cc

```diff
@@ -340,13 +340,17 @@ void callbackEndOfAfterComponentGetInput(EnergyPlusState state, void (*f)(Energy
     EnergyPlus::PluginManagement::registerNewCallback(*thisState, EnergyPlus::EMSManager::EMSCallFrom::ComponentGetInput, f);
 }
 
-// void callbackUserDefinedComponentModel(EnergyPlusState state, std::function<void ()> f) {
-//    EnergyPlus::PluginManagement::registerNewCallback(EnergyPlus::DataGlobals::emsCallFromUserDefinedComponentModel, f);
-//}
-//
-// void callbackUserDefinedComponentModel(EnergyPlusState state, void (*f)()) {
-//    callbackUserDefinedComponentModel(std::function<void ()>(f));
-//}
+void callbackUserDefinedComponentModel(EnergyPlusState state, std::function<void(EnergyPlusState)> const &f, const char *programNameInInputFile)
+{
+    auto *thisState = reinterpret_cast<EnergyPlus::EnergyPlusData *>(state); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+    EnergyPlus::PluginManagement::registerUserDefinedCallback(*thisState, f, programNameInInputFile);
+}
+
+void callbackUserDefinedComponentModel(EnergyPlusState state, void (*f)(EnergyPlusState), const char *programNameInInputFile)
+{
+    auto *thisState = reinterpret_cast<EnergyPlus::EnergyPlusData *>(state); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+    EnergyPlus::PluginManagement::registerUserDefinedCallback(*thisState, f, programNameInInputFile);
+}
```

Comment:
These are the API functions that allow the user to register their callback function with a string name.  As usual, in this cc file, we provide the void(*f) and std::function overloads.